### PR TITLE
Calculated transaction cache! 

### PR DESCRIPTION
**Pull request by fluidvanadium**
By using this implementation of WalletWrite, we can use the full version of create_proposed_transactions

this means we will be supported by the full send system of LRZ. much reduces our divergence

### CODE REVIEW

File: zingolib/src/lightclient/send.rs [hunk 1]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
 
     impl LightClient {
         /// Calculates, signs and broadcasts transactions from a proposal.
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: comment seems out of date and is same as following fn

File: zingolib/src/lightclient/send.rs [hunk 2]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
...
+                                    self.wallet.transaction_context.config.accept_server_txids,
+                                ))
+                            }
+                            Err(server_err) => {
+                                return Err(BroadcastCachedTransactionsError::Broadcast(server_err))
+                            }
+                        };
+                    }
+
+                    let non_empty_serverz_txids = NonEmpty::from_vec(txids).ok_or(
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: the type describes its non-emptiness already

File: zingolib/src/lightclient/send.rs [hunk 3]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
...
+
         async fn complete_and_broadcast<NoteRef>(
             &self,
             proposal: &Proposal<zcash_primitives::transaction::fees::zip317::FeeRule, NoteRef>,
         ) -> Result<NonEmpty<TxId>, CompleteAndBroadcastError> {
-            let submission_height = self
-                .get_submission_height()
-                .await
-                .map_err(CompleteAndBroadcastError::SubmissionHeight)?;
+            dbg!("creating transactions");
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: remove dbg plz

File: zingolib/src/lightclient/send.rs [hunk 4]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
...
+            self.wallet.create_transaction(proposal).await?;
 
-            let build_result = self.wallet.build_transaction(proposal).await?;
+            dbg!("recording transactions");
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: remove dbg plz

File: zingolib/src/lightclient/send.rs [hunk 5]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
...
-                .wallet
-                .send_to_addresses_inner(
-                    build_result.transaction(),
-                    submission_height,
-                    self.get_server_uri(),
-                )
-                .await
-                .map_err(CompleteAndBroadcastError::Broadcast)
-                .map(NonEmpty::singleton);
+            dbg!("broadcasting transactions");
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: remove dbg plz

File: zingolib/src/lightclient/send.rs [hunk 6]
```diff
@@ -81,38 +111,133 @@ pub mod send_with_proposal {
...
+            let broadcast_result = self.broadcast_created_transactions().await;
 
+            dbg!("result");
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: remove dbg plz

File: zingolib/src/lightclient/send.rs [hunk 7]
```diff
@@ -181,16 +304,12 @@ pub mod send_with_proposal {
...
-                    .await
-                    .unwrap_err()
-                    .to_string(),
-            );
+            dbg!(lc
+                .complete_and_broadcast(&proposal)
+                .await
+                .unwrap_err()
+                .to_string(),);
+            todo!("refinish test");
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: unfinished

File: zingolib/src/utils.rs [hunk 8]
```diff
@@ -36,3 +36,37 @@ pub(crate) use build_method;
 pub(crate) use build_method_push;
 #[cfg(test)]
 pub(crate) use build_push_list;
+
+/// mod
```
Comment (Oscar-Pepper) [inline zingolib/src/utils.rs]: too wordy ;)

File: zingolib/src/wallet/send.rs [hunk 9]
```diff
@@ -152,8 +145,6 @@ impl LightWallet {
             return Err(BuildTransactionError::ExchangeAddressesNotSupported);
         }
 
-        let step = proposal.steps().first();
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/send.rs:155]: 320 here we come!

File: zingolib/src/wallet/tx_map/trait_walletwrite.rs [hunk 10]
```diff
@@ -0,0 +1,82 @@
+//! currently only implementing one method of WalletWrite
+
+use zcash_client_backend::data_api::WalletWrite;
+
+use super::{TxMap, TxMapTraitError};
+
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/tx_map/trait_walletwrite.rs:6]: hoorahhh! love it

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
I left a few comments but main concern here is test are failing.

amazing work!!

File: zingolib/src/lightclient/send.rs (hunk 2 above)
Comment (fluidvanadium) [reply zingolib/src/lightclient/send.rs]: fixed

File: zingolib/src/lightclient/send.rs (hunk 1 above)
Comment (fluidvanadium) [reply zingolib/src/lightclient/send.rs]: updated

**Comment by fluidvanadium** [comment]
> I left a few comments but main concern here is test are failing.
> 
> amazing work!!

everything done is this PR except the test has a todo!()

**Comment by fluidvanadium** [comment]
> I left a few comments but main concern here is test are failing.
> 
> amazing work!!

everything done is this PR except the test has a todo!()

File: zingo-status/src/confirmation_status.rs [hunk 11]
```diff
@@ -198,21 +128,26 @@ impl ConfirmationStatus {
...
+    /// assert!(!ConfirmationStatus::Mempool(10.into()).is_pending_before(&10.into()));
+    /// assert!(ConfirmationStatus::Mempool(10.into()).is_pending_before(&11.into()));
+    /// assert!(!ConfirmationStatus::Confirmed(10.into()).is_pending_before(&9.into()));
+    /// assert!(!ConfirmationStatus::Confirmed(10.into()).is_pending_before(&10.into()));
+    /// assert!(!ConfirmationStatus::Confirmed(10.into()).is_pending_before(&11.into()));
+    /// ```
+    // TODO remove 'pending' and fix spend status.
     pub fn is_pending_before(&self, comparison_height: &BlockHeight) -> bool {
         match self {
             Self::Transmitted(self_height) | Self::Mempool(self_height) => {
```
Comment (Oscar-Pepper) [inline zingo-status/src/confirmation_status.rs]: why is `Calculated` not in the same category as `Transmitted` and `Mempool` when it comes to confirmation? it strikes me that this could cause a pending bug?

File: zingo-status/src/confirmation_status.rs [hunk 12]
```diff
@@ -277,6 +193,9 @@ impl ConfirmationStatus {
 impl std::fmt::Display for ConfirmationStatus {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
+            Self::Calculated(_) => {
+                write!(f, "transmitted")
```
Comment (Oscar-Pepper) [inline zingo-status/src/confirmation_status.rs]: change to calculated

File: zingolib/src/lightclient/send.rs [hunk 13]
```diff
@@ -80,39 +115,139 @@ pub mod send_with_proposal {
     }
 
     impl LightClient {
-        /// Calculates, signs and broadcasts transactions from a proposal.
+        /// When a transaction is created, it is added to a cache. This step records all cached transactions into TransactionRecord s.
+        /// overwrites confirmation status to Calculated (not broadcast) so only call this if
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs:119]: only call this if what?

File: zingolib/src/lightclient/send.rs [hunk 14]
```diff
@@ -80,39 +115,139 @@ pub mod send_with_proposal {
...
+                .await
+                .map_err(RecordCachedTransactionsError::Height)?;
+            let mut transactions_to_record = vec![];
+            if let Some(spending_data) = tx_map.spending_data_mut() {
+                for raw_tx in spending_data.cached_raw_transactions().values() {
+                    transactions_to_record.push(Transaction::read(
+                        &raw_tx[..],
+                        zcash_primitives::consensus::BranchId::for_height(
+                            &self.wallet.transaction_context.config.chain,
+                            current_height,
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: I believe this should be current height + 1 (mempool height) right?

File: zingolib/src/lightclient/send.rs [hunk 15]
```diff
@@ -80,39 +115,139 @@ pub mod send_with_proposal {
...
+                ));
+            }
+            drop(tx_map);
+            let mut txids = vec![];
+            for transaction in transactions_to_record {
+                self.wallet
+                    .transaction_context
+                    .scan_full_tx(
+                        &transaction,
+                        ConfirmationStatus::Calculated(current_height),
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: current height + 1?

File: zingolib/src/lightclient/send.rs [hunk 16]
```diff
@@ -80,39 +115,139 @@ pub mod send_with_proposal {
...
+                            .join(" & ")
+                    },
+                ))
                 .await;
 
-            result
+            let broadcast_txids = NonEmpty::from_vec(broadcast_result?)
+                .ok_or(CompleteAndBroadcastError::EmptyList)?;
+
+            Ok(dbg!(broadcast_txids))
```
Comment (Oscar-Pepper) [inline zingolib/src/lightclient/send.rs]: remove dbg!

File: zingolib/src/lightclient/send.rs (hunk 7 above)
Comment (Oscar-Pepper) [reply zingolib/src/lightclient/send.rs]: still unfinished :D

File: zingolib/src/testutils/chain_generics/fixtures.rs [hunk 17]
```diff
@@ -514,14 +514,6 @@ where
...
     let tertiary = environment.create_client().await;
     let expected_fee = fee_tables::one_to_one(shpool, pool, true);
-    // assert_eq!(
-    //     secondary
-    //         .propose_send_all(tertiary,
-    //         get_base_address(tertiary, pool))
-    //         .await
-    //         .into_u64(),
-    //     0
-    // );
```
Comment (Oscar-Pepper) [inline zingolib/src/testutils/chain_generics/fixtures.rs:524]: this needs uncommenting and fixing

File: zingolib/src/wallet/transaction_context.rs [hunk 18]
```diff
@@ -103,6 +103,7 @@ mod decrypt_transaction {
             block_time: Option<u32>, // block_time should only be None when re-scanning a tx that already exists in the wallet
             price: Option<f64>,
         ) {
+            dbg!("scanning transaction txid {} ", transaction.txid());
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/transaction_context.rs]: dbg!!!!

File: zingolib/src/wallet/transaction_records_by_id.rs [hunk 19]
```diff
@@ -116,18 +116,14 @@ impl TransactionRecordsById {
     }
     /// Invalidates all transactions from a given height including the block with block height `reorg_height`
     ///
-    /// All information above a certain height is invalidated during a reorg.
+    /// All information above a certain height is invalidated during a reorg
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/transaction_records_by_id.rs]: i am in favor of full stops in doc comments. non-blocking, its your call

File: zingolib/src/wallet/transaction_records_by_id.rs [hunk 20]
```diff
@@ -116,18 +116,14 @@ impl TransactionRecordsById {
...
             .values()
             .filter_map(|transaction_metadata| {
-                if transaction_metadata
-                    .status
-                    .is_confirmed_after_or_at(&reorg_height)
-                    || transaction_metadata
-                        .status
-                        .is_pending_after_or_at(&reorg_height)
+                // doesnt matter the status: if it happen after a reorg, eliminate it
+                if transaction_metadata.status.get_height() >= reorg_height
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/transaction_records_by_id.rs:126]: as the comment suggests below, we should only be invalidating confirmed txs. otherwise a re-org could cause Calculated / Transmitted but unsuccessful transactions to be discarded. anyway this is the same behaviour as before so this is non-blocking. and we havent implemented read/write for the cache anyway which would be best implemented by adding Transaction to TransactionRecord, lets leave this until its more relavent

File: zingolib/src/wallet/transaction_records_by_id.rs [hunk 21]
```diff
@@ -371,6 +367,7 @@ impl TransactionRecordsById {
             .iter()
             .for_each(|t| println!("Removing expired mempool tx {}", t));
 
+        // doesnt actually remove the TransactionRecord. all this does is change the spend status to not spent. this is broken because SpendStatus contains a duplicate source of truth as to the ConfirmationStatus of the spending transaction.
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/transaction_records_by_id.rs]: this comment is wrong. it does also delete the record. we will do work in the future to decouple spend status but for now its stable

File: zingolib/src/wallet/tx_map/trait_walletread.rs [hunk 22]
```diff
@@ -384,13 +384,13 @@ mod tests {
     #[test]
     fn get_target_and_anchor_heights_err() {
         let transaction_records_and_maybe_trees = TxMap::new_treeless_address_free();
-        assert_eq!(
+        assert!(matches!(
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/tx_map/trait_walletread.rs:387]: i discovered a `assert_matches!` macro.. go nuts!

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
great work! left some requests

File: zingolib/src/wallet/tx_map/trait_walletread.rs (hunk 22 above)
Comment (Oscar-Pepper) [reply zingolib/src/wallet/tx_map/trait_walletread.rs:387]: oh its unstable... boooo

**Review by Oscar-Pepper — APPROVED** [review]
